managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedLedgerImpl.java

```diff
@@ -2369,33 +2375,52 @@ void internalTrimLedgers(boolean isTruncate, CompletableFuture<?> promise) {
             // skip ledger if retention constraint met
             for (LedgerInfo ls : ledgers.headMap(slowestReaderLedgerId, false).values()) {
-                boolean expired = hasLedgerRetentionExpired(ls.getTimestamp());
-                boolean overRetentionQuota = isLedgerRetentionOverSizeQuota();
-
-                if (log.isDebugEnabled()) {
-                    log.debug(
-                            "[{}] Checking ledger {} -- time-old: {} sec -- "
-                                    + "expired: {} -- over-quota: {} -- current-ledger: {}",
-                            name, ls.getLedgerId(), (clock.millis() - ls.getTimestamp()) / 1000.0, expired,
-                            overRetentionQuota, currentLedger.getId());
-                }
+                // currentLedger can not be deleted
                 if (ls.getLedgerId() == currentLedger.getId()) {
                     log.debug("[{}] Ledger {} skipped for deletion as it is currently being written to", name,
                             ls.getLedgerId());
                     break;
-                } else if (expired || isTruncate) {
-                    log.debug("[{}] Ledger {} has expired or be truncated, expired is {}, isTruncate is {}, ts {}", name, ls.getLedgerId(), expired,  isTruncate, ls.getTimestamp());
+                }
+                // if truncate, all ledgers besides currentLedger are going to be deleted
+                if (isTruncate){
+                    log.debug("[{}] Ledger {} has been truncated, , ts {}", name, ls.getLedgerId(), ls.getTimestamp());
                     ledgersToDelete.add(ls);
-                } else if (overRetentionQuota || isTruncate) {
-                    log.debug("[{}] Ledger {} is over quota or be truncated, overRetentionQuota is {}, isTruncate is {}", name, ls.getLedgerId(), overRetentionQuota, isTruncate);
+                    continue;
+                }
+
+                // if neither currentLedger nor truncate
+                if (!retentionSizeQuotaMet) {
+                    totalSizeToDelete += ls.getSize();
+                    boolean overRetentionQuota = isLedgerRetentionOverSizeQuota();
+                    boolean overRetentionQuotaAfterDelete = isLedgerRetentionOverSizeQuotaAfterDelete(totalSizeToDelete);
+                    if (overRetentionQuota) {
+                        if (overRetentionQuotaAfterDelete) {
+                            log.debug("[{}] Ledger {} is over quota", name, ls.getLedgerId());
+                            ledgersToDelete.add(ls);
+                            continue;
+                        } else {
+                            retentionSizeQuotaMet = true;
+                            ledgersToDelete.add(ls);
+                            continue;
+                        }
+                    } else {
+                        retentionSizeQuotaMet = true;
+                    }
+                }
+
+                if (hasLedgerRetentionExpired(ls.getTimestamp())) {
```

Comment:
I'm not sure why the PR can fix the problem. Looks the problem is the existing logic is if the ledger over retention quota or expired, the ledger will be added into the `ledgersToDelete`, the correct logic is if the ledger over retention quota and expired, we should add it into the `ledgersToDelete` right?

But looks like the new change still adds the ledger to the `ledgersToDelete` after over retention quota without checking the expired.

Reply:
>  the correct logic is if the ledger over retention quota and expired, we should add it into the ledgersToDelete right?

I think the logic of  retention policy means  either over retention quota or expired, the ledger should be delete.

Reply:
Oh. sorry, My mistake.